### PR TITLE
Handle block comments with text on the same line as <# #>

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
## PR Summary

Fixes https://github.com/PowerShell/vscode-powershell/issues/2845

This will better support block comments with stuff on the line:

```
<# foo
bar
baz #>
```

and also make sure that saving doesn't mess up the style it previously used.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] PR has tests
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

File: src/features/PowerShellNotebooks.ts [hunk 1]
```diff
@@ -91,13 +92,31 @@ class PowerShellNotebookContentProvider implements vscode.NotebookContentProvide
...
+        let openBlockCommentOnOwnLine: boolean = false;
+
         // Iterate through all lines in a document (aka ps1 file) and group the lines
         // into cells (markdown or code) that will be rendered in Notebook mode.
         // tslint:disable-next-line: prefer-for-of
         for (let i = 0; i < lines.length; i++) {
             // Handle block comments
             if (insideBlockComment) {
-                if (lines[i] === "#>") {
+                if (lines[i].endsWith("#>")) {
```
Comment (rjmholt) [inline src/features/PowerShellNotebooks.ts:105]: What if the line is something like:

```powershell
<# Hello #> Get-Process
```

Or

```powershell
Get-Process <# Here's how we do this #>
```

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: None of that is gonna work for reasons before this change.

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: Do you mean this change doesn't intend to fix those cases because there's some other reason why those cases won't work?

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: Right so the middle two "just work" because they're gonna show up in a code cell.

The first one is broken because currently I don't handle `<#` and `#>` being on the same line.

The last one is broken and I'm not really sure how to handle it while also not changing the file being saved.

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: It looks like the first two are particularly pernicious because those lines pass the `endsWith` checks in one of the conditions.

Is there a reason we have to attack it as lines rather than like this:

```typescript
let index = -1;
const codeCells = [];
const commentCells = [];
while (true) {
    const commentStart = content.indexOf('<#', index + 1);
    if (commentStart === -1) {
        break;
    }

    let commentEnd = content.indexOf('#>, commentStart + 2); // '#>'.length
    if (commentEnd === -1) {
        commentEnd === content.length - 1;
    }

    codeCells.push(content.substring(index, commentStart));
    codeCells.push(content.substring(commentStart, commentEnd));
    index = commentEnd + 2;
}
```

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: > It looks like the first two are particularly pernicious

no it's only the first one. The `#>` check is only if `insideBlockComment`

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: the bigger problem is saving though, not reading in as cells.

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: As when cells are modified and we need to save them back to the script?

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: Yeah exactly. That's the hard part because now you have to keep track of when the comment stops and where the code begins...

and also this:

```pwsh
Get-ChildItem <# look at this sick comment
it's amazing
#> <# oh shit now what happens #> Get-Foo <# 
oh god not another
#>
```

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: Something like:

```typescript
let content = '';
for (let i = 0; i < codeCells.length; i++) {
    content += codeCells[i];
    content += commentCells[i];
   // There are various edge cases where the cell counts are different
}
```

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: > and also this:

In that one I imagine there's three code cells, the middle one being empty

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: > Something like:

What is this snippet trying to show?

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: > What is this snippet trying to show?

How the file saves. Just a matter of putting the cells back together into the script

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: but you don't have that info, you only have _cells_. https://github.com/PowerShell/vscode-powershell/blob/master/src/features/PowerShellNotebooks.ts#L225-L249

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: maybe in the metadata we can store if they end in a newline...

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: Well if we don't assume cells are broken up by newline, we wouldn't need to insert them again, no?

The cell itself would contain the newline (i.e. the code cells would start with newlines). Perhaps that's an issue though...

Also I notice we currently mangle the newlines in the file -- it might be worth changing that for Windows users

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: > Well if we don't assume cells are broken up by newline, we wouldn't need to insert them again, no?

I think that would make this:

```pwsh
<#
asdf
#>
gci
<#
foo
#>
```

have a cell in Notebook mode look like:

```pwsh
[blank newline]
gci
```

which is awkward.

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: > Also I notice we currently mangle the newlines in the file -- it might be worth changing that for Windows users

that's fixed in this PR

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: There's the other part of the equation:

```
# Foo
# Bar
# Baz
```

^ That should be a markdown cell.

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: > which is awkward.

Basically it sounds like block comments don't actually align well with the cell concept because they don't separate code segments in a straightforward way. They can begin and end on the same line, and many users use them for things like documenting individual parameters.

Cells seem to intrinsically (and intuitively) demand the notion of starting on a new line, which is something a line comment but not a block comment guarantees.

But given that people will, and probably should, expect block comments to work to some extent in notebooks, I would imagine the following scenarios:

```powershell
Get-Thing

    <# Separator #>

Get-Thing2
```

```powershell
Get-Thing

<# junk
  # =================
  junk #>

Get-Thing2
```

Where the block comment is in some way surrounded by newlines, it's a cell separator. One question there is whether we ignore spaces before or after the comment delineator.

Otherwise, block comments appear in the code cell blocks.

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: But certainly all the scenarios in the earlier comment would appear in the code block, and not create new cells

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: Any reason we can't say "the start of the block comment needs to start the line, and the end of the block comment needs to end a line"?

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: I think that's reasonable yeah. So it's as simple as the `startsWith` and `endsWith` tests, and the only edge case is where they're on the same line

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:105]: Yeah, and for now, I'd like to defer that until someone opens an issue. It increases complexity because now "on save" I need to add `<#` and `#>`.

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:105]: Honestly, it's going to come up

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:123]: Last commit has support for `<# foo #>`

Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:123]: but we live by the constraint that:

> your block comment must start a line with `<#` and end a line with `#>`.

File: src/features/PowerShellNotebooks.ts [hunk 2]
```diff
@@ -228,23 +291,36 @@ class PowerShellNotebookContentProvider implements vscode.NotebookContentProvide
         const retArr: string[] = [];
         for (const cell of document.cells) {
             if (cell.cellKind === vscode.CellKind.Code) {
-                retArr.push(...cell.document.getText().split(/\r|\n|\r\n/));
+                retArr.push(...cell.document.getText().split(/\r\n|\r|\n/));
```
Comment (rjmholt) [inline src/features/PowerShellNotebooks.ts:294]: Is the `\r` by itself needed? Seems like a weird character to hit, but something we shouldn't care about -- if people want to embed random carriage returns in their file, should we care (given that we're not on Mac OS 8)?

File: src/features/PowerShellNotebooks.ts [hunk 3]
```diff
@@ -228,23 +291,36 @@ class PowerShellNotebookContentProvider implements vscode.NotebookContentProvide
...
+                    }
                 } else {
-                    retArr.push(...cell.document.getText().split(/\r|\n|\r\n/).map((line) => `# ${line}`));
+                    retArr.push(...cell.document.getText().split(/\r\n|\r|\n/).map((line) => `# ${line}`));
                 }
             }
         }
 
-        await vscode.workspace.fs.writeFile(targetResource, new TextEncoder().encode(retArr.join("\n")));
+        await vscode.workspace.fs.writeFile(targetResource, new TextEncoder().encode(retArr.join(EOL)));
```
Comment (rjmholt) [inline src/features/PowerShellNotebooks.ts:323]: This will replace newlines in a script when it changes operating systems. For example, if we have a script checked in to git and you edit it, then I edit it, I'll change all the newlines in it. Harder to implement an alternative, but something we should aim not to do -- I wouldn't want my text editor silently reencoding line endings as a side-effect of an unrelated operation. I think worth documenting and/or opening a new issue for

File: src/features/PowerShellNotebooks.ts (hunk 2 above)
Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts:294]: Yeah I guess we don't care about Mac OS 8 :)

Comment (rjmholt) [reply src/features/PowerShellNotebooks.ts:294]: Oops there was a MacOS 9! Well that one too

File: src/features/PowerShellNotebooks.ts (hunk 3 above)
Comment (TylerLeonhardt) [reply src/features/PowerShellNotebooks.ts]: So to solve this, I included metadata at the document level that says which line ending that file uses

File: src/features/PowerShellNotebooks.ts [hunk 4]
```diff
@@ -78,7 +77,7 @@ function CreateCell(cellKind: vscode.CellKind, source: string[], metadata: IPowe
         cellKind,
         language: cellKind === vscode.CellKind.Markdown ? "markdown" : "powershell",
         outputs: [],
-        source: source.join(EOL),
+        source: source.join("\n"),
```
Comment (TylerLeonhardt) [inline src/features/PowerShellNotebooks.ts:80]: This is only used in the _rendering_ of notebooks so `\n` is totally fine.

**Comment by TylerLeonhardt** [comment]
@rjmholt I believe this is good to go.
